python/ray/data/datasource/file_based_datasource.py

```diff
@@ -248,7 +248,7 @@ def read_files(
                     partitions = parse(read_path)
 
                 with RetryingContextManager(
-                    self._open_input_source(fs, read_path, **open_stream_args),
+                    lambda: self._open_input_source(fs, read_path, **open_stream_args),
```

Comment:
Hmm should fs here be a `RetryingPyFileSystem`, and should this retry behavior be handled by that?

 https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/file_based_datasource.py#L324

Reply:
Would this be better:

```
fs: RetryingPyFileSystem = RetryingPyFileSystem(fs, data_context=...)
with fs.open_input_stream(...) as f:  # retries with the DataContext params
    ...
```

Reply:
+1.

Assuming `filesystem` is a `RetryingPyFileSystem`, I'm actually not sure if we need a `RetryingContextManager`. 

Do we know if `RetryingPyFileSystem` is being used here? If not, I think we should probably fix that first (and hopefully it'd also resolve this issue)